pyproject.toml

```diff
@@ -29,18 +29,18 @@ classifiers = [
 dynamic = ["version", "readme"]
 
 dependencies = [
-    "botocore >=1.34.70, <1.34.163", # NOTE: When updating, always keep `project.optional-dependencies` aligned
+    "botocore >=1.35.0, <1.35.4", # NOTE: When updating, always keep `project.optional-dependencies` aligned
     "aiohttp >=3.9.2, <4.0.0",
     "wrapt >=1.10.10, <2.0.0",
     "aioitertools >=0.5.1, <1.0.0",
 ]
 
 [project.optional-dependencies]
 awscli = [
-    "awscli >=1.32.70, <1.33.45",
```

Comment:
I see this was getting out of sync in the past. Is it really necessary to cap it below a certain patch version?

Reply:
You raise a valid point there. I cannot explain why these constraints are needed.

@thehesiod: Can you inform us? Wouldn't `awscli` and `boto3` be sufficiently constrained by their own dependency on `botocore`, without us having to keep this in sync?

Reply:
the problem is pip has historically not obeyed secondary dependencies.  So for example, if you already have awscli installed, and then ask to install botocore[awscli] and there's no version range specified it won't complain.

Reply:
Pip has a dependency resolver capable of backtracking for the past 4–5 years. It shouldn't be a problem today.

Reply:
lets test before removing three scenarios
1. install w/o awscli installed
2. install w/ too old awscli installed
3. install w/ too new awscli installed

i bet at least one of these will still fail

Reply:
I'm involved with the packaging ecosystem a lot and I can tell you right away that resolvelib is built in a way that it cannot output the resolved set of deps that are in conflict, taking into account what's installed already.

Reply:
I still don't understand enough to weigh in on this one. Shall I open a new issue specific to this?

Reply:
```
$ pip3 list | grep -E '(pip|awscli|boto)'
awscli                    1.32.51
boto3                     1.34.51
botocore                  1.34.131
pip                       24.1.2
pipenv                    2023.12.1
$ cat setup.py | grep awscli
# NOTE: When updating botocore make sure to update awscli/boto3 versions below
#     'awscli': ['awscli>=1.32.70,<1.33.14'],
12:15 PM alexmohr@amohr-TYQNW6P9KQ ~/dev/third_party/aiobotocore ((HEAD detached at 23ef769))$ cat aiobotocore/__init__.py 
__version__ = '2.13.2'
$ pip3 install dist/aiobotocore-2.13.2.tar.gz 
Looking in indexes: https://pypi.fbn.org/simple
Processing ./dist/aiobotocore-2.13.2.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: botocore<1.34.132,>=1.34.70 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (1.34.131)
Requirement already satisfied: aiohttp<4.0.0,>=3.9.2 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (3.9.2)
Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (1.15.0)
Requirement already satisfied: aioitertools<1.0.0,>=0.5.1 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (0.11.0)
Requirement already satisfied: aiosignal>=1.1.2 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.3.1)
Requirement already satisfied: attrs>=17.3.0 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (23.1.0)
Requirement already satisfied: frozenlist>=1.1.1 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.4.0)
Requirement already satisfied: multidict<7.0,>=4.5 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (6.0.4)
Requirement already satisfied: yarl<2.0,>=1.0 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.9.2)
Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (0.10.0)
Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (2.8.2)
Requirement already satisfied: urllib3!=2.2.0,<3,>=1.25.4 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (1.26.16)
Requirement already satisfied: six>=1.5 in /opt/homebrew/lib/python3.11/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (1.16.0)
Requirement already satisfied: idna>=2.0 in /opt/homebrew/lib/python3.11/site-packages (from yarl<2.0,>=1.0->aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (3.4)
Building wheels for collected packages: aiobotocore
  Building wheel for aiobotocore (pyproject.toml) ... done
  Created wheel for aiobotocore: filename=aiobotocore-2.13.2-py3-none-any.whl size=77125 sha256=ab6eafb110f2a2575ce1114d2c534ba6355d53a063a6e90c56290a0120f38374
  Stored in directory: /Users/alexmohr/Library/Caches/pip/wheels/f0/d5/9b/7721c8451a6477fa27d10a677bdc26794ca1b21378af3686e4
Successfully built aiobotocore
Installing collected packages: aiobotocore
Successfully installed aiobotocore-2.13.2

[notice] A new release of pip is available: 24.1.2 -> 24.2
[notice] To update, run: python3.11 -m pip install --upgrade pip
```
no complaints

Reply:
fyi my setuptools was `71.1.0`

Reply:
even after upgrading:
```bash
$ pip3 list | grep -E '(pip|setuptools|awscli|boto)'
awscli                    1.32.51
boto3                     1.34.51
botocore                  1.34.131
pip                       24.2
pipenv                    2023.12.1
setuptools                73.0.1
$ pip3 install dist/aiobotocore-2.13.2.tar.gz 
Looking in indexes: https://pypi.fbn.org/simple
Processing ./dist/aiobotocore-2.13.2.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: botocore<1.34.132,>=1.34.70 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (1.34.131)
Requirement already satisfied: aiohttp<4.0.0,>=3.9.2 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (3.9.2)
Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (1.15.0)
Requirement already satisfied: aioitertools<1.0.0,>=0.5.1 in /opt/homebrew/lib/python3.11/site-packages (from aiobotocore==2.13.2) (0.11.0)
Requirement already satisfied: aiosignal>=1.1.2 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.3.1)
Requirement already satisfied: attrs>=17.3.0 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (23.1.0)
Requirement already satisfied: frozenlist>=1.1.1 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.4.0)
Requirement already satisfied: multidict<7.0,>=4.5 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (6.0.4)
Requirement already satisfied: yarl<2.0,>=1.0 in /opt/homebrew/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (1.9.2)
Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (0.10.0)
Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (2.8.2)
Requirement already satisfied: urllib3!=2.2.0,<3,>=1.25.4 in /opt/homebrew/lib/python3.11/site-packages (from botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (1.26.16)
Requirement already satisfied: six>=1.5 in /opt/homebrew/lib/python3.11/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.34.132,>=1.34.70->aiobotocore==2.13.2) (1.16.0)
Requirement already satisfied: idna>=2.0 in /opt/homebrew/lib/python3.11/site-packages (from yarl<2.0,>=1.0->aiohttp<4.0.0,>=3.9.2->aiobotocore==2.13.2) (3.4)
Building wheels for collected packages: aiobotocore
  Building wheel for aiobotocore (pyproject.toml) ... done
  Created wheel for aiobotocore: filename=aiobotocore-2.13.2-py3-none-any.whl size=77152 sha256=520068bb46d241d144071b1de76a4667e5b11b1084ea782fef008c34fac9ff40
  Stored in directory: /Users/alexmohr/Library/Caches/pip/wheels/f0/d5/9b/7721c8451a6477fa27d10a677bdc26794ca1b21378af3686e4
Successfully built aiobotocore
Installing collected packages: aiobotocore
Successfully installed aiobotocore-2.13.2
12:32 PM alexmohr@amohr-TYQNW6P9KQ $ pip3 list | grep awscli
awscli                    1.32.51
```

Reply:
I even tried with extras `'awscli': [],` and `pip3 install dist/aiobotocore-2.13.2.tar.gz[awscli]`, didn't work either

Reply:
setuptools is a build backend. It has nothing to do with installing. The important bit is pip or any other frontend. Also, it doesn't really make sense to use sdists if you have wheels since they will still be built as an intermediate step regardless.

To verify if your venv is consistent, use the `pip check` command.

Reply:
Also, the version of setuptools in the target venv is irrelevant. It's not even shipped with CPython anymore.
The build is happening in an isolated ephemeral env where it's installed from `[build-system].requires`. Assuming you used the latest code base where it's declared and modern pip, it won't go the legacy route and will use PEP 517.

Reply:
> 2.13.2

Oh, so maybe you're building an older version of the lib.

Reply:
Anyway, the most indicative information would be metadata if the deps you're trying to constrain.

Reply:
yes, however this is why we have the version req in the extras, it works correctly if you have it without having to rely on something like pipenv/poetry/etc

Reply:
perhaps we should put this in a [discussion](https://github.com/aio-libs/aiobotocore/discussions)? :)

Reply:
this is mostly a shortcut for workflows that want to do a simple install of aiobotocore along with compatible versions of awscli/boto3/etc w/o having to resort to using a tool like pipenv/poetry/etc to do recursive algo to find compatible versions of all three

Reply:
Considering all that information, I believe that relaxing the constraint or removing the extras would both be considered breaking changes and should be decided and performed separately from this PR. Feel free to open a separate discussion or issue. There might even be input from the wider community with scenarios we haven't thought of.